autoload/phpactor.vim

```diff
@@ -459,49 +459,13 @@ endfunction
 " RPC -->-->-->-->-->--
 """""""""""""""""""""""
 
-function! s:searchDirectoryUpwardForRootPatterns(initialDirectory, workspaceRootPatterns, fallbackDirectory)
-  if index(g:phpactorGlobalRootPatterns, '/') < 0
-    call add(g:phpactorGlobalRootPatterns, '/')
-  endif
-
-  let l:directory = a:initialDirectory
-
-  while index(g:phpactorGlobalRootPatterns, l:directory) < 0
-    if s:directoryMatchesToPatterns(l:directory, a:workspaceRootPatterns)
-      return l:directory
-    endif
-
-    let l:directory = fnamemodify(l:directory, ':h')
-  endwhile
-
-  if index(g:phpactorGlobalRootPatterns, l:directory) >= 0
-    let l:directory = a:fallbackDirectory
-  endif
-
-  return l:directory
-endfunction
-
-function s:directoryMatchesToPatterns(directory, patterns) abort
-  for l:pattern in a:patterns
-    if (filereadable(a:directory .'/'. l:pattern))
-      return v:true
-    endif
-  endfor
-
-  return v:false
-endfunction
-
 function! phpactor#rpc(action, arguments)
     " Remove any existing output in the message window
     execute ':redraw'
 
     let request = { "action": a:action, "parameters": a:arguments }
 
-    let l:workspaceDir = s:searchDirectoryUpwardForRootPatterns(
-          \ fnamemodify(phpactor#_path(), ':h'),
-          \ g:phpactorProjectRootPatterns,
-          \ g:phpactorInitialCwd
-          \)
+    let l:workspaceDir = (get(b:, 'phpactorProject', v:null) isnot v:null) ? b:phpactorProject.primaryRootPath : g:phpactorInitialCwd
```

Comment:
This will solve the issue #985
But I think we can do even better, right it only makes possible to use phpactor root detection of the directory from where vim was open.

It should not be phpactor responsibility to resolve the directory in the first place, we should only ensure that we provide the root directory when sending a request.

I mentioned it because on this line I think we should consider the all root detection mechanism as external.
It would reduce the complexity by getting back to real need: provide a root directory.
Initially phpactor was using `g:phpactorInitialCwd`, to prevent any BC break we need to keep this variable with it's value being the directory from which vim was started.
From this point we want to be able to provide a dynamic directory depending on the buffer, meaning we will need to "ask" each buffer what is its root directory.
We can translate this need by
> We want to be able to apply different strategies at runtime

So instead of coupling this piece of code to any system, let's couple it to an abstract strategy that could be defined by the anyone.
```suggestion
    let l:workspaceDir = phpactor#getRootDirectory()
```

With
```vim
if !exists('g:PhpactorRootDirectoryStrategy')
  let g:PhpactorRootDirectoryStrategy = {-> g:phpactorInitialCwd}
endif

function! phpactor#getRootDirectory() abort
  let l:Strategy = get(b:, 'PhpactorRootDirectoryStrategy', g:PhpactorRootDirectoryStrategy)

  if type(function('type')) != type(l:Strategy)
    let l:Strategy = {-> g:phpactorInitialCwd}
  endif

  return l:Strategy()
endfunction
```

At this point the user is free to define both global and buffer local strategies.
We don't have any BC break since the default behavior is the same as before the first attempt to resolve the root directory.
Example of configuration:

#### To use with your system
```vim
" vimrc
let g:PhpactorRootDirectoryStrategy = {-> get(get(b:, 'phpactorProject', {}), 'primaryRootPath', g:phpactorInitialCwd)}
```

#### To use vim pwd (for users with [vim-rooter](https://github.com/airblade/vim-rooter) for instance)
```vim
" vimrc
let g:PhpactorRootDirectoryStrategy = function('getcwd')
```

The goal of this proposal is not to diminish your work at all, it's simply to provide a way of integrating in a extensible way.
So thanks for all what you have done anyone can use phpactor out of box with multiple projects in one vim instances.
Users having already a setup handling this notion can plug into phpactor to reuse what they have.
Users with no need for it will keep things as they were.

Reply:
I kept such idea (without any implementation) as a backup in case of non-acceptance the whole work, but it should be the default approach - not backup. Your proposition sounds reasonable: gives an abstraction that contains a minimal implementation and can be replaced by any implementation maintained independently in some separate project. 

The whole implementation of this approach is yours so please make PR that closes this after merge. 

@dantleech @elythyr thanks for a lot of critical word.

Reply:
Thanks so much for all your work so far :) Even if this PR isn't merged as-is there is lots of valuable stuff here which we can take forward (especially with the Vader tests).

I've found it quite difficult to follow everything going on here, but I think we could do something like:

- Allowing project-root strategy to be selected (custom or phpactor CWD)
- Strategy for changing CWD if the user opens up a file != the filesystem branch of the current project root (with confirmation).

Either as part of this PR or new PR(s) - would that cover everything?

Reply:
Let this PR die in the current form and (before) implement @elythyr's proposition in a separate one. 

@elythyr would you like to make such PR? I can do it just as well.

Reply:
You can do it if you want.
The ideal would be to do it quickly enough so it will be in the next release.

Let me know if you won't have the time.

Reply:
It can be unpredictable so we can assume I have not.